analyzers/src/SonarAnalyzer.CSharp/Rules/NamedPlaceholdersShouldBeUnique.cs

```diff
@@ -0,0 +1,122 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class NamedPlaceholdersShouldBeUnique : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S6677";
+    private const string MessageFormat = "Message template placeholder '{0}' is not unique.";
+
+    private static readonly HashSet<string> MicrosoftExtensionsLoggingMethods =
+        [
+            "Log",
+            "LogCritical",
+            "LogDebug",
+            "LogError",
+            "LogInformation",
+            "LogTrace",
+            "LogWarning"
+        ];
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+        ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+            {
+                if (c.Node is InvocationExpressionSyntax invocation
+                    && TemplateArgument(invocation, c.SemanticModel) is { } argument
+                    && MessageTemplates.Parse(argument.Expression.ToString()) is { Success: true } result)
+                {
+                    RaiseAtDuplicates(c, argument, result.Placeholders);
+                }
+            },
+            SyntaxKind.InvocationExpression);
+
+    private static void RaiseAtDuplicates(SonarSyntaxNodeReportingContext context, ArgumentSyntax templateArgument, MessageTemplates.Placeholder[] placeholders)
+    {
+        var duplicates = placeholders
+            .Where(x => !int.TryParse(x.Name, out var _)) // exclude index placeholders like {42}
+            .GroupBy(x => x.Name)
+            .Where(x => x.Count() > 1).Select(x => x.Skip(1))
+            .SelectMany(x => x);
+
+        foreach (var placeholder in duplicates)
+        {
+            var spanStart = templateArgument.Expression.GetLocation().SourceSpan.Start + placeholder.Start;
+            var location = Location.Create(context.Tree, new(spanStart, placeholder.Length));
+            context.ReportIssue(Diagnostic.Create(Rule, location, placeholder.Name));
+        }
+    }
+
+    private ArgumentSyntax TemplateArgument(InvocationExpressionSyntax invocation, SemanticModel model) =>
+        MicrosoftTemplateArgument(invocation, model)
+        ?? SerilogTemplateArgument(invocation, model)
+        ?? NLogTemplateArgument(invocation, model);
+
+    private ArgumentSyntax MicrosoftTemplateArgument(InvocationExpressionSyntax invocation, SemanticModel model)
+    {
+        if (MicrosoftExtensionsLoggingMethods.Contains(invocation.GetIdentifier().ToString())
+            && invocation.ArgumentList.Arguments is { Count: >= 2 }
```

Comment:
This check serves two purposes:
- We cannot be out-of-bounds later inside `MessageArgument`
- It's a performance short-circuit boost:

The usual case is going to be `logger.LogXXX(template,...)`. If the arguments are one, it means there are no placeholders, and it will look like `logger.LogXXX(template)` so we shouldn't bother parsing this. If it indeed has placeholders but the user did not supply them after the template, they will get [CA2017](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2017) anyways.